udata_hydra/crawl/process_check_data.py

```diff
@@ -31,10 +31,36 @@ async def process_check_data(check_data: dict) -> Tuple[Record, bool]:
 
     is_first_check: bool = last_check is None
 
+    # Calculate next check date
+    if is_first_check:
+        next_check = datetime.now(timezone.utc) + timedelta(hours=config.CHECK_DELAY_DEFAULT)
+
+    else:
+        if has_check_changed:
+            # Resource has been modified since last check, we will check it again in in the soonest of the CHECK_DELAYS
+            next_check = datetime.now(timezone.utc) + timedelta(
+                hours=config.CHECK_DELAYS[0]
+            )  # TODO: should this be CHECK_DELAY_DEFAULT?
+        else:
+            # Resource has not been modified since last check:
+            # if the time since last check is greater than the longest delay in CHECK_DELAYS, next check will be after the longest delay
+            # if the time since last check is less than CHECK_DELAYS[i], next check will be after this CHECK_DELAYS[i]
+            previous_delay: timedelta = datetime.now(timezone.utc) - datetime.fromisoformat(
+                last_check["created_at"]
+            )
+            if previous_delay > timedelta(hours=config.CHECK_DELAYS[-1]):
+                next_check = datetime.now(timezone.utc) + timedelta(hours=config.CHECK_DELAYS[-1])
+            else:
+                for d in config.CHECK_DELAYS:
+                    if previous_delay <= timedelta(hours=d):
+                        next_check = datetime.now(timezone.utc) + timedelta(hours=d)
+                        break
+        check_data["next_check"] = next_check
+
```

Comment:
I had a hard time understanding what that was doing, maybe it's just me...

Would that be clearer like so (I may be missing some detail or edge cases):

```suggestion
            # Resource has not been modified since last check: 
            # Re-check it after a delay just greater than the time elapsed since the last check.
            since_last_check: timedelta = datetime.now(timezone.utc) - datetime.fromisoformat(
                last_check["created_at"]
            )
            now: datetime = datetime.now(timezone.utc)
            # At most, re-check after the greatest delay.
            next_check = now + timedelta(hours=config.CHECK_DELAYS[-1])
            # Should we re-check it sooner? Loop through the CHECK_DELAYS from longest to shortest.
            for delay in reversed(config.CHECK_DELAYS):
                delay_delta: timedelta = timedelta(hours=delay)
                if since_last_check <= delay_delta:
                    next_check = now + delay_delta
                else:
                    # Previous delay was the last one greater than `since_last_check`, keep it.
                    break
        check_data["next_check"] = next_check
```

I realize my rewrite is slightly longer, it feels clearer to me, but it could very well be because i just (re)wrote it, feel free to simply dismiss this.

Reply:
Very nice, thanks, although I wonder if I prefer going from the shorter delay to the longest... As the shortest should be the default (more safe), and going to longer delays should be considered as "fallbacks", if that make sense.

I might make a mix your solution with mine though!